### PR TITLE
Fix error 10022 in Worker.WriteOutput() caused by RioSocket

### PR DESCRIPTION
**Pull request by hebinhuang**
There are random failures on Worker.WriteOutput() with socket error 10022 (An invalid argument was supplied). The reason of the error is because the RioSocketWrapper pass the managed buffer to native buffer and call native code to send the data too many. 

This change is to modify SocketStream.Write with a send data cache, so that to reduce the frequency of calling Riosocket send operation.

### CODE REVIEW

**Comment by skaarthik** [comment]
@hebinhuang - let's validate this fix using the socket streaming as well as FastSML scenarios.

**Comment by hebinhuang** [comment]
This change have fixed the following bugs:
1. Modify SocketStream.Write with a send data cache, so that to reduce the frequency of calling Riosocket send operation and improve performance.
2. Adjust ByteBufChunk size to reduce the frequency of register/unregister RIO buffer, to fix the error 10022 during socket I/O operation (It looks like there is a bug in Windows RIO APIs - the high frequency of Register/unregister RIO buffer caused the error 10022 for socket I/O operation).
3. Separate the background threads to dequeue Send and Receive results. Before this change, we use a single background thread to dequeue Send and Receive results.

This change have been tested on local and yarn cluster mode.

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs [hunk 1]
```diff
@@ -21,6 +22,21 @@ internal sealed class ByteBufPool
...
+
+        /// <summary>
+        /// The default chunk order used to calculate chunk size and ensure it is a multiple of a segment size.
+        /// </summary>
+        public const int DefaultChunkOrder = 12; // 65536 << 12 = 256 MBytes per chunk
+
+        /// <summary>
+        /// The trial count to try allocation.
+        /// </summary>
+        private const int TrialsCount = 1000;
```
Comment (skaarthik) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs]: Need more context on this constant as code comments - purpose, reason behind the magic number etc.

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs [hunk 2]
```diff
@@ -110,23 +116,31 @@ public static ByteBufPool UnsafeDefault
...
             {
-                return byteBuf;
-            }
+                if (q050.Allocate(out byteBuf) || q025.Allocate(out byteBuf) ||
+                    q000.Allocate(out byteBuf) || qInit.Allocate(out byteBuf) ||
+                    q075.Allocate(out byteBuf))
+                {
+                    return byteBuf;
+                }
+                Thread.Sleep(0);
```
Comment (skaarthik) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs]: why is this needed?

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/DefaultSocketWrapper.cs [hunk 3]
```diff
@@ -80,6 +83,28 @@ public Stream GetStream()
         }
 
         /// <summary>
+        /// Returns a stream used to receive data only.
+        /// </summary>
+        /// <returns>The underlying Stream instance that be used to receive data</returns>
+        public Stream GetInputStream()
+        {
+            var readBufferSize = int.Parse(Environment.GetEnvironmentVariable(ConfigurationService.CSharpWorkerReadBufferSizeEnvName) ?? "8192");
```
Comment (skaarthik) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/DefaultSocketWrapper.cs]: Need context on the magic number used here

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/ISocketWrapper.cs [hunk 4]
```diff
@@ -38,6 +38,18 @@ internal interface ISocketWrapper : IDisposable
         Stream GetStream();
```
Comment (skaarthik) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/ISocketWrapper.cs]: do we still need this method given that you have added methods for getting input and output stream separately?

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs [hunk 5]
```diff
@@ -128,25 +129,34 @@ private void Init()
             var maxThreads = 1;
             if (useThreadPool)
             {
-                maxThreads = Environment.ProcessorCount;
+                maxThreads = (Environment.ProcessorCount + 1) >> 1;
```
Comment (skaarthik) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs]: why?

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/SocketStream.cs [hunk 6]
```diff
@@ -87,10 +88,29 @@ public override long Position
         }
 
         /// <summary>
-        /// Flushes data from the stream.  This is meaningless for us, so it does nothing.
+        /// Flushes data in send cache to the stream.
         /// </summary>
         public override void Flush()
         {
+            if (sendDataCache != null && sendDataCache.IsReadable())
```
Comment (skaarthik) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/SocketStream.cs]: do we need to handle the case when IsReadable is false?

File: csharp/Adapter/Microsoft.Spark.CSharp/Proxy/Ipc/StreamingContextIpcProxy.cs [hunk 7]
```diff
@@ -306,7 +305,8 @@ private void ProcessCallbackRequest(object socket)
                                 logger.LogDebug("receive close cmd from Scala side");
                                 break;
                             }
-                            else if (cmd == "callback")
```
Comment (skaarthik) [inline csharp/Adapter/Microsoft.Spark.CSharp/Proxy/Ipc/StreamingContextIpcProxy.cs]: what is the reason for this change?

File: csharp/Worker/Microsoft.Spark.CSharp/Worker.cs [hunk 8]
```diff
@@ -343,12 +343,20 @@ private static IFormatter ProcessCommand(Stream inputStream, Stream outputStream
...
+                    {
+                        logger.LogError("WriteOutput() failed at iteration {0}", count);
+                        throw;
+                    }
                     count++;
                     funcProcessWatch.Start();
                 }
 
-                logger.LogDebug("Output entries count: " + count);
+                logger.LogInfo("Output entries count: " + count);
```
Comment (skaarthik) [inline csharp/Worker/Microsoft.Spark.CSharp/Worker.cs]: what is the reason for this change?

**Comment by skaarthik** [comment]
@hebinhuang - code cov number is dropping in this PR. Please address that.

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs (hunk 2 above)
Comment (hebinhuang) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs]: Thread.Sleep(0) issues a pause instruction on each loop iteration, and not fall back to a more expensive wait on.

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs (hunk 1 above)
Comment (hebinhuang) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs]: Added

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/SocketStream.cs (hunk 6 above)
Comment (hebinhuang) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/SocketStream.cs]: No. If IsReadable is false, it means no data need to be send out.

File: csharp/Adapter/Microsoft.Spark.CSharp/Proxy/Ipc/StreamingContextIpcProxy.cs (hunk 7 above)
Comment (hebinhuang) [inline csharp/Adapter/Microsoft.Spark.CSharp/Proxy/Ipc/StreamingContextIpcProxy.cs]: else if (cmd == "callback") is redundant code. I got the suggestion from ReSharp tool.

File: csharp/Worker/Microsoft.Spark.CSharp/Worker.cs (hunk 8 above)
Comment (hebinhuang) [inline csharp/Worker/Microsoft.Spark.CSharp/Worker.cs]: Prompt the log level here, which give us information about the output entries. By default, LogDebug is not shown in the log.

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs (hunk 5 above)
Comment (hebinhuang) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs]: We separate the work threads to process send and receive completion at the below code. So just make the total number of threads still equals to the machine's processor count.

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/ISocketWrapper.cs (hunk 4 above)
Comment (hebinhuang) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/ISocketWrapper.cs]: Tao Wang introduced the input and output stream at the CSharpWorker only. I do not do lots of changes on drive side. So still keep the function here.

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/DefaultSocketWrapper.cs (hunk 3 above)
Comment (hebinhuang) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/DefaultSocketWrapper.cs]: This one I moved from Tao wang's changes. But I have added the comments.

**Comment by hebinhuang** [comment]
@skaarthik , the total number is dropped by 0.34%. The reason is that this PR introduced a new Completion Queue for send completion in native code, and added error handler for the new Completion Queue. So the decreased number is from the error handler in native code.

**Comment by skaarthik** [comment]
@hebinhuang - lines 235 - 240 in SocketStream.cs are newly added code and not covered by any tests. BTW, with pure unit tests, even error handling code can be covered easily.

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs (hunk 2 above)
Comment (skaarthik) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs]: I guess the question is why do we need a pause instruction? Can you add your response as code comments also?

**Comment by hebinhuang** [comment]
The lines you mentioned are already covered by TestRioSocket in SocketWrapperTest.

See the coverage https://codecov.io/github/Microsoft/Mobius/commit/0fe576daad96b2318aabf41127a67c88326e4f3d, the code have been covered.

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs (hunk 5 above)
Comment (skaarthik) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs]: Do we know how this work in YARN?

File: csharp/Worker/Microsoft.Spark.CSharp/Worker.cs (hunk 8 above)
Comment (skaarthik) [inline csharp/Worker/Microsoft.Spark.CSharp/Worker.cs]: Right, but how is this "info" useful other than for debugging?

**Comment by skaarthik** [comment]
@hebinhuang - those lines are not covered. Look at the snapshot from CodeCov below

image

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs (hunk 2 above)
Comment (hebinhuang) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs]: Added.

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs (hunk 5 above)
Comment (hebinhuang) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs]: These code is only for Multithread Worker. In simple worker, there are only two threads (separately for send and receive completion). No difference between local and yarn cluster mode.

Comment (skaarthik) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs]: How will this calculation for maxThreads work within a YARN container where it is possible to control the number of cores?

File: csharp/Worker/Microsoft.Spark.CSharp/Worker.cs (hunk 8 above)
Comment (hebinhuang) [inline csharp/Worker/Microsoft.Spark.CSharp/Worker.cs]: this "info" give us a sense of total entries be processed by a worker, just like the "info" for "func process time". Based on the "info" of "Output entries count" and "func process time", we can have the idea of processing time of each process.
So I still think it is better to use "info" than "Debug".

**Comment by hebinhuang** [comment]
Now, it is covered by last merge.

image

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs (hunk 5 above)
Comment (hebinhuang) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs]: Just print out the processor number from Yarn cluster:
CSharpWorker [11500] is running on a machine with 32 processors

In Multithread worker mode, if YARN arrange multiple executors into one machine, it will start multiple CsharpWorker - so the total threads will be "processors X csharpworkers".

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufChunk.cs [hunk 9]
```diff
@@ -286,6 +288,7 @@ private void Dispose(bool disposing)
                 return;
             }
 
+            logger.LogDebug("Disposing ByteBufChunk [{0}].", RuntimeHelpers.GetHashCode(this).ToString("X"));
```
Comment (qualiu) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufChunk.cs]: The HashCode will not change not only in all the threads of the process, but also the processes every time started.
So, if you want to add a log line and print extra info, maybe it's more useful to print the object other than the invariant HashCode.

**Comment by qualiu** [comment]
LGTM

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufChunk.cs (hunk 9 above)
Comment (hebinhuang) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufChunk.cs]: Changed to the below:

logger.LogDebug("Disposing ByteBufChunk [{0}].", ToString());

File: csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs (hunk 5 above)
Comment (hebinhuang) [inline csharp/Adapter/Microsoft.Spark.CSharp/Network/RioNative.cs]: As your mentioned, we can get the executor cores in cluster/standalone mode.

CSharpWorker [6260] is started with cores: 28

**Comment by hebinhuang** [comment]
@skaarthik , I have update the PR.

**Comment by skaarthik** [comment]
LGTM
